### PR TITLE
Always allow "no-other-choice" pre-release dependencies when resolving collection dependency tree

### LINKED ISSUE #79168: Unable to install collection with type: git
### Summary

I'm trying to install some collections using `ansible-galaxy install -r requirements.yml` but am unable to install a specific version from a git repository.

### Issue Type

Bug Report

### Component Name

galaxy

### Ansible Version

```console
$ ansible --version
ansible [core 2.13.5]
  config file = /Users/josh/Code/server/ansible.cfg
  configured module search path = ['/Users/josh/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/josh/Code/server/hapy/venv/lib/python3.10/site-packages/ansible
  ansible collection location = /Users/josh/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/josh/Code/server/hapy/venv/bin/ansible
  python version = 3.10.8 (main, Oct 13 2022, 09:48:40) [Clang 14.0.0 (clang-1400.0.29.102)]
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
DEFAULT_HOST_LIST(/Users/josh/Code/server/ansible.cfg) = ['/Users/josh/Code/server/ansible/inventory_aws_ec2.yml']
DISPLAY_SKIPPED_HOSTS(/Users/josh/Code/server/ansible.cfg) = False
HOST_KEY_CHECKING(/Users/josh/Code/server/ansible.cfg) = False
INTERPRETER_PYTHON(/Users/josh/Code/server/ansible.cfg) = python

CALLBACK:
========

default:
_______
display_skipped_hosts(/Users/josh/Code/server/ansible.cfg) = False
pretty_results(/Users/josh/Code/server/ansible.cfg) = True

minimal:
_______
pretty_results(/Users/josh/Code/server/ansible.cfg) = True

CONNECTION:
==========

paramiko_ssh:
____________
host_key_checking(/Users/josh/Code/server/ansible.cfg) = False
ssh_args(/Users/josh/Code/server/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=60s -o ControlPath=/tmp/ansible-ssh-%h-%p-%r -o ForwardAgent=yes

ssh:
___
host_key_checking(/Users/josh/Code/server/ansible.cfg) = False
ssh_args(/Users/josh/Code/server/ansible.cfg) = -o ControlMaster=auto -o ControlPersist=60s -o ControlPath=/tmp/ansible-ssh-%h-%p-%r -o ForwardAgent=yes
```

### OS / Environment

macOS 12.5

### Steps to Reproduce

requirements.yml
```yaml (paste below)
collections:
  - name: https://github.com/ansible-collections/amazon.aws.git
    type: git
    version: 395d4fda1ec549020d070d4465835166094e4777
  - name: community.aws
    version: 3.5.0

```

### Expected Results

I expect the collection to be installed at the specified git commit tag with no issues. 

### Actual Results

```console
% ansible-galaxy install -r ansible/requirements.yml 
Starting galaxy collection install process
Process install dependency map
Cloning into '/Users/josh/.ansible/tmp/ansible-local-44490z57uazrk/tmpztxrx5cz/amazon.aws48wn3_f9'...
remote: Enumerating objects: 18877, done.
remote: Counting objects: 100% (2652/2652), done.
remote: Compressing objects: 100% (799/799), done.
remote: Total 18877 (delta 2008), reused 2345 (delta 1810), pack-reused 16225
Receiving objects: 100% (18877/18877), 15.95 MiB | 4.49 MiB/s, done.
Resolving deltas: 100% (14006/14006), done.
Note: switching to '395d4fda1ec549020d070d4465835166094e4777'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 395d4fda ec2_instance: fix to handle create instance in specified AZ (#1150)
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* amazon.aws:>=3.4.0 (dependency of community.aws:3.5.0)
* amazon.aws:6.0.0-dev0 (dependency of "virtual collection Git repo")
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #80048: Galaxy collection install dependency fails mysteriously
### Summary

I recently uploaded my first collection to Ansible Galaxy 🎉

Unfortunately, I am experiencing some strange behavior which I believe is related to `requirements.yml`.

`ansible-galaxy collection install o0_o.host` results in:

```
Starting galaxy collection install process
Process install dependency map
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* o0_o.host:* (direct request)
```

It appears to be trying to install itself as a dependency, but this is not reflected in my `requirements.yml` file.

```
---
roles:
  - name: o0_o.inventory
collections:
  - name: ansible.netcommon
  - name: ansible.posix
  - name: community.general
  - name: community.routeros
  - name: community.network
```

Most confusingly, specifying either version or installing from a tarball works as expected.

Each of the following will complete without an error:

```
ansible-galaxy collection install o0_o.host:1.0.0-alpha.1
```

```
ansible-galaxy collection install o0_o.host:1.0.0-alpha.2
```

```
ansible-galaxy collection install o0_o-host-1.0.0-alpha.2.tar.gz
```

```
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Downloading https://galaxy.ansible.com/download/o0_o-host-1.0.0-alpha.2.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/o0_o-host-1.0.0-alpha.2-k8q8lx0u
Installing 'o0_o.host:1.0.0-alpha.2' to '/Users/o0-o/.ansible/collections/ansible_collections/o0_o/host'
Downloading https://galaxy.ansible.com/download/ansible-posix-1.5.1.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-posix-1.5.1-aeh7481o
o0_o.host:1.0.0-alpha.2 was installed successfully
Installing 'ansible.posix:1.5.1' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/posix'
Downloading https://galaxy.ansible.com/download/community-general-6.3.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-general-6.3.0-dsdjucim
ansible.posix:1.5.1 was installed successfully
Installing 'community.general:6.3.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/general'
Downloading https://galaxy.ansible.com/download/community-network-5.0.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-network-5.0.0-8unm3c6c
community.general:6.3.0 was installed successfully
Installing 'community.network:5.0.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/network'
Downloading https://galaxy.ansible.com/download/ansible-netcommon-4.1.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-netcommon-4.1.0-8z9q_2qb
community.network:5.0.0 was installed successfully
Installing 'ansible.netcommon:4.1.0' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/netcommon'
Downloading https://galaxy.ansible.com/download/ansible-utils-2.9.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-utils-2.9.0-lr2u1tmo
ansible.netcommon:4.1.0 was installed successfully
Installing 'ansible.utils:2.9.0' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/utils'
Downloading https://galaxy.ansible.com/download/community-routeros-2.7.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-routeros-2.7.0-bvkoq41m
ansible.utils:2.9.0 was installed successfully
Installing 'community.routeros:2.7.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/routeros'
community.routeros:2.7.0 was installed successfully
```

### Issue Type

Bug Report

### Component Name

requirements.yml

### Ansible Version

```console
$ ansible --version
ansible [core 2.14.2]
  config file = None
  configured module search path = ['/Users/o0-o/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/7.2.0/libexec/lib/python3.11/site-packages/ansible
  ansible collection location = /Users/o0-o/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.11.2 (main, Feb 16 2023, 03:07:35) [Clang 14.0.0 (clang-1400.0.29.202)] (/usr/local/Cellar/ansible/7.2.0/libexec/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
```

### OS / Environment

macOS

### Steps to Reproduce

```
ansible-galaxy collection install o0_o.host
```

### Expected Results

It should install successfully with dependencies.

```
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Downloading https://galaxy.ansible.com/download/o0_o-host-1.0.0-alpha.2.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/o0_o-host-1.0.0-alpha.2-k8q8lx0u
Installing 'o0_o.host:1.0.0-alpha.2' to '/Users/o0-o/.ansible/collections/ansible_collections/o0_o/host'
Downloading https://galaxy.ansible.com/download/ansible-posix-1.5.1.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-posix-1.5.1-aeh7481o
o0_o.host:1.0.0-alpha.2 was installed successfully
Installing 'ansible.posix:1.5.1' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/posix'
Downloading https://galaxy.ansible.com/download/community-general-6.3.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-general-6.3.0-dsdjucim
ansible.posix:1.5.1 was installed successfully
Installing 'community.general:6.3.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/general'
Downloading https://galaxy.ansible.com/download/community-network-5.0.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-network-5.0.0-8unm3c6c
community.general:6.3.0 was installed successfully
Installing 'community.network:5.0.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/network'
Downloading https://galaxy.ansible.com/download/ansible-netcommon-4.1.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-netcommon-4.1.0-8z9q_2qb
community.network:5.0.0 was installed successfully
Installing 'ansible.netcommon:4.1.0' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/netcommon'
Downloading https://galaxy.ansible.com/download/ansible-utils-2.9.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/ansible-utils-2.9.0-lr2u1tmo
ansible.netcommon:4.1.0 was installed successfully
Installing 'ansible.utils:2.9.0' to '/Users/o0-o/.ansible/collections/ansible_collections/ansible/utils'
Downloading https://galaxy.ansible.com/download/community-routeros-2.7.0.tar.gz to /Users/o0-o/.ansible/tmp/ansible-local-11511pdq6gyg0/tmprccg_5ng/community-routeros-2.7.0-bvkoq41m
ansible.utils:2.9.0 was installed successfully
Installing 'community.routeros:2.7.0' to '/Users/o0-o/.ansible/collections/ansible_collections/community/routeros'
community.routeros:2.7.0 was installed successfully
```

### Actual Results

```console
Starting galaxy collection install process
Process install dependency map
ERROR! Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* o0_o.host:* (direct request)
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by webknjaz**
Prior to this patch, when `--pre` CLI flag was not passed, the
dependency resolver would treat concrete collection dependency
candidates (Git repositories, subdirs, tarball URLs, or local dirs or
files etc) as not meeting the requirements.

This patch makes it so pre-releases in any concrete artifact
references, and the ones being specifically pinned dependencies or
user requests, met anywhere in the dependency tree, are allowed
unconditionally.

This is achieved by moving the pre-release check from
`is_satisfied_by()` to the `find_matches()` hook, following the
Pip's example.

As a bonus, this change also fixes the situation when a collection
pre-releases weren't considered if it didn't have any stable releases.
This now works even if `--pre` wasn't requested explicitly.

Ref #73416
Ref #79112
Fixes #79168
Fixes #80048
Resolves #81605

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

##### SUMMARY

$sbj.

##### ISSUE TYPE

- Bugfix Pull Request

##### ADDITIONAL INFORMATION

Special thanks to @pradyunsg for hinting how to process pre-releases in the resolvelib provider correctly.

### CODE REVIEW

**Comment by webknjaz** [comment]
@s-hertel I borrowed your tests from #81605 but haven't added any extra ones. I think more might be needed.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 1]
```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
...
-            requirement.ver.startswith('<') or
-            requirement.ver.startswith('>') or
-            requirement.ver.startswith('!=')
-        ) or self._is_user_requested(candidate)
+        # NOTE: Finally, if a concrete artifact is met, like a Git repository,
+        # NOTE: subdirs, a tarball URL, or a local dir or file etc,
+        # NOTE: pre-releases are allowed regardless of any other circumstances.
+        allow_pre_release = (
+            self._with_pre_releases or
+            candidate.is_concrete_artifact or not (
```
Comment (s-hertel) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: I think this is okay, though it is much broader than `type: git`, and we don't have much test coverage for dependencies that are concrete artifacts as it pertains to pre-releases. At least we know the missing test coverage/potential issues of the change is limited to dependencies, since direct reqs were (mostly) handled by the last `or`.

If this does have any unintended side effects, the only workaround would probably be `--no-deps`.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 2]
```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
...
+                requirement.ver == '*' or
+                requirement.ver.startswith('<') or
+                requirement.ver.startswith('>') or
+                requirement.ver.startswith('!=')
+            ) or self._is_user_requested(candidate)
```
Comment (s-hertel) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: I think we should clean up the (now) unused stuff that was added in https://github.com/ansible/ansible/pull/73416/files, and rename `_is_user_requested` method to something more accurate since it's now just handling a `type: galaxy` corner case.

**Review by s-hertel — COMMENTED** [review]
Does the pre-release hint https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/__init__.py#L1817 need a qualification since `--pre` will have no bearing on concrete artifacts (direct or indirect)?

**Comment by webknjaz** [comment]
> Does the pre-release hint [`devel`/lib/ansible/galaxy/collection/__init__.py#L1817](https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection/__init__.py?rgh-link-date=2023-08-31T13%3A39%3A32Z#L1817) need a qualification since `--pre` will have no bearing on concrete artifacts (direct or indirect)?

I'm not sure what you're suggesting. Do mean that we should change the message to say “except for concrete artifacts”? Could you elaborate?

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: I meant for this to apply beyond Git as while making this patch, I recognized that the problem would manifest in each of those cases otherwise. So I've gone for a generic solution. But you're right, we should have better coverage.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Yeah, maybe. I didn't look that far, just wanted to demonstrate the minimum generic change necessary.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 1 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: I was questioning the additional change on top of that regarding dependencies (as the lay-user will see them). The generic fix is a bit redundant, but once the dead code is removed the simplification definitely seems like a net positive if it doesn't have side effects.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (s-hertel) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: It's a small change, but I am less confident about backwards compatibility.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: @s-hertel looks like that entire commit can be reverted :thinking:

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py:372]: Reverted. Marking this as resolved.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Reverted. Marking this as resolved.

**Comment by webknjaz** [comment]
> how about saying it it only applies to requirements from galaxy servers, since that would be more meaningful to a user.

That's a great idea!

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Had to revert it back. It seems to still be necessary.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: I don't think it's dead. It just covers a different part of the corner case.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Thinking more, we might need that `is_pinned` property after all. To facilitate removal of `_is_user_requested`.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: > @s-hertel looks like that entire commit can be reverted 🤔

UPD: So while working on this, just reverting the commit didn't work, which sent me off into the rabbit hole. I realized that we have a deeper problem with how we process pre-releases — if a requirement on any level of depth in the dependency tree has a pinned dependency that wan't taken into account. With the help of my pip maintainer  friend, we've confirmed that pip does not have the same problem — if a pinned dependency is a pre-release, it's allowed automatically in pip, without requiring a `--pre` which I regard as the correct behavior. With that, _I knew_ what bug we had but couldn't make up what's the difference with how pip does things. Luckily, Pradyun pointed out the pip's debugging reporter and I used it to figure out what the resolver does. He also made a comment along the lines of processing pre-releases in a different hook — `find_matches()`, not `is_satisfied_by()`! This all led took me to the right exploration path and I think I've finally figured it out: _we need to stop disregarding pinned versions from any requirements on the current round that are pinned!_

It looks like I nailed it, finally. This PR needs some love before getting out of Draft but it feels close to being ready.

@s-hertel feel free to run some testing of your own.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 3]
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
                 for version, _none_src_server in coll_versions
             ]
 
+        def requirement_is_pinned(requirement):
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: _TODO:_ Move this to a requirement method.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 4]
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
...
             tmp_candidate = Candidate(fqcn, version, src_server, 'galaxy', None)
 
             unsatisfied = False
             for requirement in requirements:
                 unsatisfied |= not self.is_satisfied_by(requirement, tmp_candidate)
+                unsatisfied |= (
+                    # NOTE: Only allow pre-release candidates if we want pre-releases
+                    # NOTE: or the req ver was an exact match with the pre-release
+                    # NOTE: version. Another case where we'd want to allow
+                    # NOTE: pre-releases is when there are several user requirements
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: _TODO:_ This doesn't need clarification anymore as it's now more generic.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 5]
```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
...
+            return not (
+                requirement.ver == '*' or
+                requirement.ver.startswith('<') or
+                requirement.ver.startswith('>') or
+                requirement.ver.startswith('!=')
+            ) or requirement.ver[0].isdigit()
+
         latest_matches = []
         signatures = []
         extra_signature_sources = []  # type: list[str]
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py:297]: _TODO:_ Figure out why this isn't reset on each outer loop cycle below.

File: test/integration/targets/ansible-galaxy-collection/tasks/main.yml [hunk 6]
```diff
@@ -134,6 +134,10 @@
   loop_control:
     loop_var: resolvelib_version
 
+- name: test installing prereleases via scm direct requests
```
Comment (webknjaz) [inline test/integration/targets/ansible-galaxy-collection/tasks/main.yml:144]: _TODO:_ Include a dependency tree truly nested transitive pinned dependencies.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 3 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Moved. Marking as resolved.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 4 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Updated. Marking as resolved.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 5 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py:297]: @s-hertel any ideas?

File: lib/ansible/galaxy/collection/__init__.py [hunk 7]
```diff
@@ -1817,15 +1817,15 @@ def _resolve_depenency_map(
         elif not req.specifier.contains(RESOLVELIB_VERSION.vstring):
             raise AnsibleError(f"ansible-galaxy requires {req.name}{req.specifier}")
 
-    if allow_pre_release:
-        pre_release_hint = ''
-    else:
-        pre_release_hint = 'Hint: Pre-releases are not installed by default unless the specific version is given. To enable pre-releases, use --pre.'
+    pre_release_hint = '' if allow_pre_release else (
+        'Hint: Pre-releases hosted on Galaxy or Automation Hub are not '
+        'installed by default unless the version is given. To enable '
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/__init__.py]: _FIXME:_ "specific" got lost

File: test/integration/targets/ansible-galaxy-collection/tasks/main.yml (hunk 6 above)
Comment (webknjaz) [reply test/integration/targets/ansible-galaxy-collection/tasks/main.yml:144]: Done. Marking as resolved.

File: lib/ansible/galaxy/collection/__init__.py (hunk 7 above)
Comment (webknjaz) [reply lib/ansible/galaxy/collection/__init__.py]: Fixed. Marking as resolved.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Alright.. After more testing and tweaks, I mark this thread as resolved.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: This and more corner cases have been addressed. Marking this thread as resolved.

**Comment by webknjaz** [comment]
@s-hertel I'm happy with how the patch turned out, it's now ready to be reviewed. If you have more test ideas, please send them my way.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 5 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py:297]: @s-hertel this seems like a bug but I chose not to attempt addressing it within this PR. Still, it'd be good to figure it out...

File: changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+bugfixes:
+- >-
+  ansible-galaxy - started allowing the use of pre-releases
+  for dependencies on any level of the dependency tree that
+  specifically demand exact pre-release versions of
+  collections and not version ranges.
+  (https://github.com/ansible/ansible/pull/81606)
+- >-
+  ansible-galaxy - started allowing the use of pre-releases
```
Comment (jborean93) [inline changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml:11]: Does pip work like this?

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py [hunk 9]
```diff
@@ -564,6 +564,15 @@ def is_concrete_artifact(self):
     def is_online_index_pointer(self):
         return not self.is_concrete_artifact
 
+    @property
+    def is_pinned(self):
+        """Indicate if the version set is considered pinned."""
```
Comment (jborean93) [inline lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Might be good to expand on what `pinned` actually means in this context.

**Review by jborean93 — COMMENTED** [review]
Just some minor comments, the rest looks good to me.

File: changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml (hunk 8 above)
Comment (webknjaz) [reply changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml:11]: Yep

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py (hunk 9 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Great suggestion, will do!

File: changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml (hunk 8 above)
Comment (webknjaz) [reply changelogs/fragments/81606-ansible-galaxy-collection-pre-releases.yml:11]: Retrieved some context for history:
* pip's rationale (which I'd apparently thumbed up years ago): https://github.com/pypa/pip/issues/5503#issuecomment-397426044
* PEP 440: https://peps.python.org/pep-0440/#handling-of-pre-releases

While googling I've also found that another tool — pipenv — might have a problem with this, but not pip itself:
* https://github.com/pypa/pipenv/issues/2022#issuecomment-382706536
* https://github.com/psf/black/issues/2115

File: lib/ansible/galaxy/dependency_resolution/dataclasses.py (hunk 9 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/dataclasses.py]: Addressed in https://github.com/ansible/ansible/pull/81606/commits/e683693fd6a2eea1ce6283dd0b7613316ff74e32.

Marking the thread as resolved.

**Review by jborean93 — APPROVED** [review]

**Comment by webknjaz** [comment]
Note for context: @s-hertel and I agreed in a private discussion on Slack, that any further improvements would go into follow-up PRs.

**Comment by webknjaz** [comment]
Backports PRs:
* https://github.com/ansible/ansible/pull/81746
* https://github.com/ansible/ansible/pull/81747
* https://github.com/ansible/ansible/pull/81748
* https://github.com/ansible/ansible/pull/81749
